### PR TITLE
[APB-2353][AO] Allow starting mapping journey with any of allowed non-MTD enrolments

### PR DESCRIPTION
**Pull request by arturopala**
Do not require IR-SA-AGENT nor HMCE-VAT-AGNT enrolment as a prerequisite to starting mapping journey.

### CODE REVIEW

File: app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala [hunk 1]
```diff
@@ -68,53 +66,48 @@ class MappingController @Inject()(override val messagesApi: MessagesApi,
   }
 
   def startSubmit: Action[AnyContent] = Action.async { implicit request =>
-    withAuthorisedAgent() {
-      implicit request: AgentRequest[AnyContent] =>
-        successful(Redirect(routes.MappingController.showAddCode()))
-    }
+    withAuthorisedAsAgent {
+      successful(Redirect(routes.MappingController.showAddCode()))
+    }()
```
[the remark this answers is not in the document]
Comment (arturopala) [reply app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala]: this actually part of a valid application of `withAuthorisedAsAgent ` which takes 2 sets of arguments

File: it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala [hunk 2]
```diff
@@ -4,19 +4,18 @@ import play.api.http.Writeable
...
 class MappingControllerISpec extends BaseControllerISpec with AuthStubs {
 
   def callEndpointWith[A: Writeable](request:Request[A]): Result = await(play.api.test.Helpers.route(app, request).get)
 
   "context root" should {
-    "redirect to start page" in {
+    "redirect to the start page" in {
       val request = FakeRequest(GET,"/agent-mapping/")
       val result = callEndpointWith(request)
       status(result) shouldBe 303
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala]: redirectLocation shouldBe
will check where you been redirected too, and that its 303

File: it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala [hunk 3]
```diff
@@ -37,42 +36,31 @@ class MappingControllerISpec extends BaseControllerISpec with AuthStubs {
...
-      val request = fakeRequest(GET, "/agent-mapping/start-submit")
-      val result = callEndpointWith(request)
-      status(result) shouldBe 303
-      redirectLocation(result).get shouldBe routes.MappingController.showAddCode().url
+    Auth.validEnrolments.foreach { serviceName =>
+      s"redirect to the add-code if the current user is logged in and has legacy agent enrolment for $serviceName" in {
+        givenAuthorisedFor(serviceName)
+        val request = fakeRequest(GET, "/agent-mapping/start-submit")
+        val result = callEndpointWith(request)
+        status(result) shouldBe 303
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala]: line 45, should check whether status is 303

Comment (arkowalski) [reply it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala]: should be built into redirectLocation

Comment (arturopala) [reply it/uk/gov/hmrc/agentmappingfrontend/controllers/MappingControllerISpec.scala]: `redirectLocation` is a standard test facility provided by `play.api.test`, it looks at the `Location` header only

File: it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala [hunk 4]
```diff
@@ -28,80 +28,62 @@ trait EndpointBehaviours extends AuthStubs {
...
     }
 
-    "redirect to the start page if the current user is logged in and does not have affinity group Agent" in {
-      givenUserIsAuthenticated(individual)
+    "redirect to the sign-in page if the current user is logged in but does not have affinity group Agent" in {
+      givenUnauthorisedWith("UnsupportedAffinityGroup")
       val request = fakeRequest(endpointMethod,endpointPath)
       val result = await(doRequest(request))
 
       result.header.status shouldBe 303
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala]: redirectLocation to replace line 36 and 37

File: it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala [hunk 5]
```diff
@@ -28,80 +28,62 @@ trait EndpointBehaviours extends AuthStubs {
...
-    }
+  protected def anEndpointReachableGivenAgentAffinityGroupAndValidEnrolment(endpointMethod: String, endpointPath: String,
+                                                                            expectCheckAgentRefCodeAudit: Boolean)
+                                                                           (doRequest: Request[AnyContentAsEmpty.type] => Result): Unit = {
+    behave like anAuthenticatedEndpoint(endpointMethod, endpointPath, doRequest)
 
-    "redirect to the start page if the current user is logged in and does not have affinity group Agent" in {
-      givenUserIsAuthenticated(individual)
+    "render the not-enrolled page if the current user is logged with affinity group Agent but is not validly enrolled" in {
+      givenUserIsAuthenticated(aStrangeAgent)
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala]: aStrangeAgent? name of this doesn't really explain much, what is strange.
you can loose the a's, and just use strangeAgent

File: it/uk/gov/hmrc/agentmappingfrontend/support/SampleUsers.scala [hunk 6]
```diff
@@ -85,8 +83,40 @@ object SampleUsers {
        |    "providerType": "GovernmentGateway"
        |  }
        |}""".stripMargin,
-    identifier = Seq(Identifier("IRAgentReference","HZ1234"))
+    activeEnrolments = Set()
   )
 
-  val individual = SampleUser("", Seq.empty, Some(InsufficientEnrolments()))
+  val anMTDAgent = SampleUser(
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentmappingfrontend/support/SampleUsers.scala]: this is not that necessary, but mtdAgent works too.

Comment (arkowalski) [reply it/uk/gov/hmrc/agentmappingfrontend/support/SampleUsers.scala]: I meant the extra character "a" you tend to do this throughout using a, an

File: app/uk/gov/hmrc/agentmappingfrontend/auth/AuthActions.scala [hunk 7]
```diff
@@ -19,78 +19,64 @@ package uk.gov.hmrc.agentmappingfrontend.auth
...
-          others.foreach { id =>
-            AuditService.auditCheckAgentRefCodeEvent(Some(id), Option(creds.providerId), Option(creds.providerType))(auditService)
-          }
-          if(activated.nonEmpty){
-            activated.foreach { id =>
-              AuditService.auditCheckAgentRefCodeEvent(Some(id), Option(creds.providerId), Option(creds.providerType))(auditService)
-            }
-            body(AgentRequest(activated, request))
+          val activeEnrolments = justAuthorisedEnrolments.enrolments.filter(_.isActivated).map(_.key)
+          val eligible = activeEnrolments.nonEmpty && activeEnrolments.intersect(Auth.validEnrolments).nonEmpty
```
[the remark this answers is not in the document]
Comment (arturopala) [reply app/uk/gov/hmrc/agentmappingfrontend/auth/AuthActions.scala]: `activeEnrolments.nonEmpty` is a fail-fast check

File: it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala (hunk 4 above)
Comment (arkowalski) [reply it/uk/gov/hmrc/agentmappingfrontend/support/EndpointBehaviours.scala]: case ResponseHeader(SEE_OTHER, headers, _) => headers.get(LOCATION)

will do these 2 checks, in 1 line

File: app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala [hunk 8]
```diff
@@ -114,3 +105,17 @@ class MappingController @Inject()(override val messagesApi: MessagesApi,
     Future successful Ok(html.not_enrolled())
   }
 }
+
+object MappingController {
+
+  val mappingForm = Form(
+    mapping(
+      "arn" -> mapping(
+        "arn" -> arn
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala]: im not sure whether the arn requires the second embedded "mapping" as the field has the same name "arn" arn"
whereas the utr has value for it.

Comment (arturopala) [reply app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala]: maybe not, but changing this is not a part of this story

**Pull request merged**